@seanm , as far as I can tell:
The thread-safety annotations are correct, but CI is red on every clang job (Android, Linux-clang, Fuzzing) for two build-config reasons:

1) clang-as-C fails to compile. The gate #if defined(__clang__) enables the real attributes for any clang build, but GUARDED_BY(sibling_member) needs C++ class-scope lookup. In C, clang errors use of undeclared identifier 'usb_devs_lock'. TSA is a C++-only feature (hence the -x c++ in the xcconfig). 
**Fix: add && defined(__cplusplus) to the gate.**

2) Fuzzing: threads_posix.h:28: fatal error: 'libusbi.h' file not found. The new #include "libusbi.h" is circular and breaks configs that don't have libusb/ on the include path (oss-fuzz uses -I.). 
**Fix: drop that include** (threads_posix.h is only ever included via libusbi.h, which already provides the macros.)

I pushed these changes here to validate my hypothesis: https://github.com/libusb/libusb/commit/df45829a89e6f71207f5323c5478892c179b465c

--- a/Xcode/common.xcconfig
+++ b/Xcode/common.xcconfig
@@ -88,6 +88,9 @@ CLANG_WARN__DUPLICATE_METHOD_MATCH = YES
 CLANG_WARN_DEPRECATED_OBJC_IMPLEMENTATIONS = YES
 GCC_WARN_UNDECLARED_SELECTOR = YES
 
+// Optionally enable thread safety warnings, which requires building as C++.
+//OTHER_CFLAGS = -Wthread-safety -x c++
+
 // Static analyzer warnings.
 CLANG_ANALYZER_NONNULL = YES
 CLANG_ANALYZER_SECURITY_FLOATLOOPCOUNTER = YES

--- a/Xcode/libusb.xcodeproj/project.pbxproj
+++ b/Xcode/libusb.xcodeproj/project.pbxproj
@@ -57,6 +57,16 @@
 		008FC0301628BC7400BC5BE2 /* listdevs.c in Sources */ = {isa = PBXBuildFile; fileRef = 008FBFE71628BA0E00BC5BE2 /* listdevs.c */; };
 		1438D77A17A2ED9F00166101 /* hotplug.c in Sources */ = {isa = PBXBuildFile; fileRef = 1438D77817A2ED9F00166101 /* hotplug.c */; };
 		1438D77F17A2F0EA00166101 /* strerror.c in Sources */ = {isa = PBXBuildFile; fileRef = 1438D77E17A2F0EA00166101 /* strerror.c */; };
+		14D9E1DF2FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E02FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E12FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E22FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E32FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E42FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E52FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E62FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E72FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
+		14D9E1E82FFD5E01005170B1 /* clang_thread_safety.h in Headers */ = {isa = PBXBuildFile; fileRef = 14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */; };
 		2018D95F24E453BA001589B2 /* events_posix.c in Sources */ = {isa = PBXBuildFile; fileRef = 2018D95E24E453BA001589B2 /* events_posix.c */; };
 		2018D96124E453D0001589B2 /* events_posix.h in Headers */ = {isa = PBXBuildFile; fileRef = 2018D96024E453D0001589B2 /* events_posix.h */; };
 		20468D70243298C100650534 /* sam3u_benchmark.c in Sources */ = {isa = PBXBuildFile; fileRef = 20468D6E243298C100650534 /* sam3u_benchmark.c */; };
@@ -279,6 +289,7 @@
 		1472E15D2B43D68600850BA3 /* macos.c */ = {isa = PBXFileReference; indentWidth = 2; lastKnownFileType = sourcecode.c.c; path = macos.c; sourceTree = "<group>"; tabWidth = 2; usesTabs = 0; };
 		1472E15F2B43D68600850BA3 /* set_option.c */ = {isa = PBXFileReference; indentWidth = 2; lastKnownFileType = sourcecode.c.c; path = set_option.c; sourceTree = "<group>"; tabWidth = 2; usesTabs = 0; };
 		1472E1602B43D69800850BA3 /* umockdev.c */ = {isa = PBXFileReference; indentWidth = 4; lastKnownFileType = sourcecode.c.c; path = umockdev.c; sourceTree = "<group>"; tabWidth = 4; usesTabs = 1; };
+		14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = clang_thread_safety.h; sourceTree = "<group>"; };
 		14EC13E12B3D5BA600CF9AD0 /* emscripten_webusb.cpp */ = {isa = PBXFileReference; indentWidth = 4; lastKnownFileType = sourcecode.cpp.cpp; path = emscripten_webusb.cpp; sourceTree = "<group>"; tabWidth = 4; usesTabs = 1; };
 		14EC13E22B3D5BBE00CF9AD0 /* sunos_usb.h */ = {isa = PBXFileReference; indentWidth = 4; lastKnownFileType = sourcecode.c.h; path = sunos_usb.h; sourceTree = "<group>"; tabWidth = 4; usesTabs = 1; };
 		14EC13E32B3D5BBE00CF9AD0 /* openbsd_usb.c */ = {isa = PBXFileReference; indentWidth = 4; lastKnownFileType = sourcecode.c.c; path = openbsd_usb.c; sourceTree = "<group>"; tabWidth = 4; usesTabs = 1; };
@@ -455,6 +466,7 @@
 				008FBF561628B7E800BC5BE2 /* io.c */,
 				008FBF5A1628B7E800BC5BE2 /* libusb.h */,
 				008FBF671628B7E800BC5BE2 /* libusbi.h */,
+				14D9E1DE2FFD5E01005170B1 /* clang_thread_safety.h */,
 				008FBF6B1628B7E800BC5BE2 /* os */,
 				1438D77E17A2F0EA00166101 /* strerror.c */,
 				008FBF7A1628B7E800BC5BE2 /* sync.c */,
@@ -561,6 +573,7 @@
 				2018D96124E453D0001589B2 /* events_posix.h in Headers */,
 				008FBF901628B7E800BC5BE2 /* libusbi.h in Headers */,
 				008FBF9B1628B7E800BC5BE2 /* threads_posix.h in Headers */,
+				14D9E1E42FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				008FBFA11628B7E800BC5BE2 /* version.h in Headers */,
 				008FBFA21628B7E800BC5BE2 /* version_nano.h in Headers */,
 			);
@@ -570,6 +583,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1E82FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F5EF26321FAA00ADF3EC /* config.h in Headers */,
 				20951C152563125200ED6351 /* libusb.h in Headers */,
 			);
@@ -579,6 +593,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1E12FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F5F026321FAA00ADF3EC /* config.h in Headers */,
 				CEA0F5F126321FAA00ADF3EC /* libusb.h in Headers */,
 			);
@@ -590,6 +605,7 @@
 			files = (
 				CEA0F5F326321FAA00ADF3EC /* config.h in Headers */,
 				20951C0625630F8F00ED6351 /* ezusb.h in Headers */,
+				14D9E1E52FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F5F426321FAA00ADF3EC /* libusb.h in Headers */,
 			);
 			runOnlyForDeploymentPostprocessing = 0;
@@ -598,6 +614,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1E02FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F5F626321FAA00ADF3EC /* config.h in Headers */,
 				CEA0F5F726321FAA00ADF3EC /* libusb.h in Headers */,
 			);
@@ -607,6 +624,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1E22FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F5F926321FAA00ADF3EC /* config.h in Headers */,
 				CEA0F5FA26321FAA00ADF3EC /* libusb.h in Headers */,
 			);
@@ -616,6 +634,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1DF2FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F5FC26321FAA00ADF3EC /* config.h in Headers */,
 				CEA0F5FD26321FAA00ADF3EC /* libusb.h in Headers */,
 			);
@@ -627,6 +646,7 @@
 			files = (
 				CEA0F5FF26321FAA00ADF3EC /* config.h in Headers */,
 				20951C0F25630FD300ED6351 /* libusb_testlib.h in Headers */,
+				14D9E1E62FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F60026321FAA00ADF3EC /* libusb.h in Headers */,
 			);
 			runOnlyForDeploymentPostprocessing = 0;
@@ -635,6 +655,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1E72FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F60226321FAA00ADF3EC /* config.h in Headers */,
 				CEA0F60326321FAA00ADF3EC /* libusb.h in Headers */,
 			);
@@ -644,6 +665,7 @@
 			isa = PBXHeadersBuildPhase;
 			buildActionMask = 2147483647;
 			files = (
+				14D9E1E32FFD5E01005170B1 /* clang_thread_safety.h in Headers */,
 				CEA0F60526321FAA00ADF3EC /* config.h in Headers */,
 				CEA0F60626321FAA00ADF3EC /* libusb.h in Headers */,
 			);

--- a/examples/dpfp.c
+++ b/examples/dpfp.c
@@ -320,7 +320,7 @@ static void LIBUSB_CALL cb_mode_changed(struct libusb_transfer *transfer)
 
 static int set_mode_async(unsigned char data)
 {
-	unsigned char *buf = malloc(LIBUSB_CONTROL_SETUP_SIZE + 1);
+	unsigned char *buf = (unsigned char *)malloc(LIBUSB_CONTROL_SETUP_SIZE + 1);
 	struct libusb_transfer *transfer;
 
 	if (!buf) {

--- a/libusb/clang_thread_safety.h
+++ b/libusb/clang_thread_safety.h
@@ -1,0 +1,93 @@
+/*
+ * Internal header for libusb using Clang's thread safety analysis feature.
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef LIBUSB_CLANG_THREAD_SAFETY_H
+#define LIBUSB_CLANG_THREAD_SAFETY_H
+
+// <https://clang.llvm.org/docs/ThreadSafetyAnalysis.html>
+// Enable thread safety attributes only with clang.
+// The attributes can be safely erased when compiling with other compilers.
+#if defined(__clang__) && !defined(SWIG) && defined(__cplusplus)
+  #define THREAD_ANNOTATION_ATTRIBUTE__(x)   __attribute__((x))
+#else
+  #define THREAD_ANNOTATION_ATTRIBUTE__(x)   // no-op
+#endif
+
+#define CAPABILITY(x) \
+  THREAD_ANNOTATION_ATTRIBUTE__(capability(x))
+
+#define SCOPED_CAPABILITY \
+  THREAD_ANNOTATION_ATTRIBUTE__(scoped_lockable)
+
+#define GUARDED_BY(x) \
+  THREAD_ANNOTATION_ATTRIBUTE__(guarded_by(x))
+
+#define PT_GUARDED_BY(x) \
+  THREAD_ANNOTATION_ATTRIBUTE__(pt_guarded_by(x))
+
+#define ACQUIRED_BEFORE(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(acquired_before(__VA_ARGS__))
+
+#define ACQUIRED_AFTER(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(acquired_after(__VA_ARGS__))
+
+#define REQUIRES(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(requires_capability(__VA_ARGS__))
+
+#define REQUIRES_SHARED(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(requires_shared_capability(__VA_ARGS__))
+
+#define ACQUIRE(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(acquire_capability(__VA_ARGS__))
+
+#define ACQUIRE_SHARED(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(acquire_shared_capability(__VA_ARGS__))
+
+#define RELEASE(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(release_capability(__VA_ARGS__))
+
+#define RELEASE_SHARED(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(release_shared_capability(__VA_ARGS__))
+
+#define RELEASE_GENERIC(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(release_generic_capability(__VA_ARGS__))
+
+#define TRY_ACQUIRE(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(try_acquire_capability(__VA_ARGS__))
+
+#define TRY_ACQUIRE_SHARED(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(try_acquire_shared_capability(__VA_ARGS__))
+
+#define EXCLUDES(...) \
+  THREAD_ANNOTATION_ATTRIBUTE__(locks_excluded(__VA_ARGS__))
+
+#define ASSERT_CAPABILITY(x) \
+  THREAD_ANNOTATION_ATTRIBUTE__(assert_capability(x))
+
+#define ASSERT_SHARED_CAPABILITY(x) \
+  THREAD_ANNOTATION_ATTRIBUTE__(assert_shared_capability(x))
+
+#define RETURN_CAPABILITY(x) \
+  THREAD_ANNOTATION_ATTRIBUTE__(lock_returned(x))
+
+#define NO_THREAD_SAFETY_ANALYSIS \
+  THREAD_ANNOTATION_ATTRIBUTE__(no_thread_safety_analysis)
+
+#endif

--- a/libusb/core.c
+++ b/libusb/core.c
@@ -55,11 +55,11 @@ static int default_context_refcnt;
 static usbi_atomic_t default_debug_level = -1;
 #endif
 static usbi_mutex_static_t default_context_lock = USBI_MUTEX_INITIALIZER;
-static struct usbi_option default_context_options[LIBUSB_OPTION_MAX];
+static struct usbi_option default_context_options[LIBUSB_OPTION_MAX] GUARDED_BY(default_context_lock);
 
 
 usbi_mutex_static_t active_contexts_lock = USBI_MUTEX_INITIALIZER;
-struct list_head active_contexts_list;
+struct list_head active_contexts_list GUARDED_BY(active_contexts_lock);
 
 /**
  * \mainpage libusb-1.0 API Reference
@@ -1557,7 +1557,7 @@ out:
 }
 
 static void do_close(struct libusb_context *ctx,
-	struct libusb_device_handle *dev_handle)
+	struct libusb_device_handle *dev_handle) EXCLUDES(ctx->flying_transfers_lock)
 {
 	struct usbi_transfer *itransfer;
 	struct usbi_transfer *tmp;
@@ -1835,7 +1835,7 @@ int API_EXPORTED libusb_set_configuration(libusb_device_handle *dev_handle,
  * \see libusb_set_auto_detach_kernel_driver()
  */
 int API_EXPORTED libusb_claim_interface(libusb_device_handle *dev_handle,
-	int interface_number)
+	int interface_number) EXCLUDES(dev_handle->lock)
 {
 	int r = 0;
 
@@ -1879,7 +1879,7 @@ out:
  * \see libusb_set_auto_detach_kernel_driver()
  */
 int API_EXPORTED libusb_release_interface(libusb_device_handle *dev_handle,
-	int interface_number)
+	int interface_number) EXCLUDES(dev_handle->lock)
 {
 	int r;
 
@@ -1924,7 +1924,7 @@ out:
  * \returns another LIBUSB_ERROR code on other failure
  */
 int API_EXPORTED libusb_set_interface_alt_setting(libusb_device_handle *dev_handle,
-	int interface_number, int alternate_setting)
+	int interface_number, int alternate_setting) EXCLUDES(dev_handle->lock)
 {
 	usbi_dbg(HANDLE_CTX(dev_handle), "interface %d altsetting %d",
 		interface_number, alternate_setting);
@@ -2184,7 +2184,7 @@ int API_EXPORTED libusb_kernel_driver_active(libusb_device_handle *dev_handle,
  * \see libusb_kernel_driver_active()
  */
 int API_EXPORTED libusb_detach_kernel_driver(libusb_device_handle *dev_handle,
-	int interface_number)
+	int interface_number) EXCLUDES(dev_handle->lock)
 {
 	usbi_dbg(HANDLE_CTX(dev_handle), "interface %d", interface_number);
 
@@ -2220,7 +2220,7 @@ int API_EXPORTED libusb_detach_kernel_driver(libusb_device_handle *dev_handle,
  * \see libusb_kernel_driver_active()
  */
 int API_EXPORTED libusb_attach_kernel_driver(libusb_device_handle *dev_handle,
-	int interface_number)
+	int interface_number) EXCLUDES(dev_handle->lock)
 {
 	usbi_dbg(HANDLE_CTX(dev_handle), "interface %d", interface_number);
 
@@ -2259,7 +2259,7 @@ int API_EXPORTED libusb_attach_kernel_driver(libusb_device_handle *dev_handle,
  * \see libusb_set_configuration()
  */
 int API_EXPORTED libusb_set_auto_detach_kernel_driver(
-	libusb_device_handle *dev_handle, int enable)
+	libusb_device_handle *dev_handle, int enable) EXCLUDES(dev_handle->lock)
 {
 	if (!(usbi_backend.caps & USBI_CAP_SUPPORTS_DETACH_KERNEL_DRIVER))
 		return LIBUSB_ERROR_NOT_SUPPORTED;

--- a/libusb/hotplug.c
+++ b/libusb/hotplug.c
@@ -151,7 +151,7 @@ int main (void) {
 #define VALID_HOTPLUG_FLAGS			\
 	(LIBUSB_HOTPLUG_ENUMERATE)
 
-void usbi_hotplug_init(struct libusb_context *ctx)
+void usbi_hotplug_init(struct libusb_context *ctx) NO_THREAD_SAFETY_ANALYSIS
 {
 	/* check for hotplug support */
 	if (!libusb_has_capability(LIBUSB_CAP_HAS_HOTPLUG))
@@ -292,7 +292,7 @@ void usbi_hotplug_notification(struct libusb_context *ctx, struct libusb_device 
 	usbi_mutex_unlock(&ctx->event_data_lock);
 }
 
-void usbi_hotplug_process(struct libusb_context *ctx, struct list_head *hotplug_msgs)
+void usbi_hotplug_process(struct libusb_context *ctx, struct list_head *hotplug_msgs) EXCLUDES(ctx->hotplug_cbs_lock)
 {
 	struct usbi_hotplug_callback *hotplug_cb, *next_cb;
 	struct usbi_hotplug_message *msg;
@@ -345,7 +345,7 @@ int API_EXPORTED libusb_hotplug_register_callback(libusb_context *ctx,
 	int events, int flags,
 	int vendor_id, int product_id, int dev_class,
 	libusb_hotplug_callback_fn cb_fn, void *user_data,
-	libusb_hotplug_callback_handle *callback_handle)
+	libusb_hotplug_callback_handle *callback_handle) EXCLUDES(ctx->hotplug_cbs_lock)
 {
 	struct usbi_hotplug_callback *hotplug_cb;
 
@@ -427,7 +427,7 @@ int API_EXPORTED libusb_hotplug_register_callback(libusb_context *ctx,
 }
 
 void API_EXPORTED libusb_hotplug_deregister_callback(libusb_context *ctx,
-	libusb_hotplug_callback_handle callback_handle)
+	libusb_hotplug_callback_handle callback_handle) EXCLUDES(ctx->hotplug_cbs_lock)
 {
 	struct usbi_hotplug_callback *hotplug_cb;
 	int deregistered = 0;
@@ -465,7 +465,7 @@ void API_EXPORTED libusb_hotplug_deregister_callback(libusb_context *ctx,
 
 DEFAULT_VISIBILITY
 void * LIBUSB_CALL libusb_hotplug_get_user_data(libusb_context *ctx,
-	libusb_hotplug_callback_handle callback_handle)
+	libusb_hotplug_callback_handle callback_handle) EXCLUDES(ctx->hotplug_cbs_lock)
 {
 	struct usbi_hotplug_callback *hotplug_cb;
 	void *user_data = NULL;

--- a/libusb/io.c
+++ b/libusb/io.c
@@ -1799,7 +1799,7 @@ void usbi_signal_transfer_completion(struct usbi_transfer *itransfer)
  * \returns 1 if the lock was not obtained (i.e. another thread holds the lock)
  * \ref libusb_mtasync
  */
-int API_EXPORTED libusb_try_lock_events(libusb_context *ctx)
+int API_EXPORTED libusb_try_lock_events(libusb_context *ctx) TRY_ACQUIRE(0, ctx->events_lock)
 {
 	int r;
 	unsigned int ru;
@@ -1842,7 +1842,7 @@ int API_EXPORTED libusb_try_lock_events(libusb_context *ctx)
  * \param ctx the context to operate on, or NULL for the default context
  * \ref libusb_mtasync
  */
-void API_EXPORTED libusb_lock_events(libusb_context *ctx)
+void API_EXPORTED libusb_lock_events(libusb_context *ctx) ACQUIRE(ctx->events_lock)
 {
 	ctx = usbi_get_context(ctx);
 	usbi_mutex_lock(&ctx->events_lock);
@@ -1857,7 +1857,7 @@ void API_EXPORTED libusb_lock_events(libusb_context *ctx)
  * \param ctx the context to operate on, or NULL for the default context
  * \ref libusb_mtasync
  */
-void API_EXPORTED libusb_unlock_events(libusb_context *ctx)
+void API_EXPORTED libusb_unlock_events(libusb_context *ctx) RELEASE(ctx->events_lock)
 {
 	ctx = usbi_get_context(ctx);
 	ctx->event_handler_active = 0;
@@ -1892,7 +1892,7 @@ void API_EXPORTED libusb_unlock_events(libusb_context *ctx)
  * \returns 0 if this thread must give up the events lock
  * \ref fullstory "Multi-threaded I/O: the full story"
  */
-int API_EXPORTED libusb_event_handling_ok(libusb_context *ctx)
+int API_EXPORTED libusb_event_handling_ok(libusb_context *ctx) REQUIRES(ctx->events_lock)
 {
 	unsigned int r;
 
@@ -1986,7 +1986,7 @@ void API_EXPORTED libusb_interrupt_event_handler(libusb_context *ctx)
  * \param ctx the context to operate on, or NULL for the default context
  * \ref libusb_mtasync
  */
-void API_EXPORTED libusb_lock_event_waiters(libusb_context *ctx)
+void API_EXPORTED libusb_lock_event_waiters(libusb_context *ctx) ACQUIRE(ctx->event_waiters_lock)
 {
 	ctx = usbi_get_context(ctx);
 	usbi_mutex_lock(&ctx->event_waiters_lock);
@@ -1997,7 +1997,7 @@ void API_EXPORTED libusb_lock_event_waiters(libusb_context *ctx)
  * \param ctx the context to operate on, or NULL for the default context
  * \ref libusb_mtasync
  */
-void API_EXPORTED libusb_unlock_event_waiters(libusb_context *ctx)
+void API_EXPORTED libusb_unlock_event_waiters(libusb_context *ctx) RELEASE(ctx->event_waiters_lock)
 {
 	ctx = usbi_get_context(ctx);
 	usbi_mutex_unlock(&ctx->event_waiters_lock);
@@ -2029,7 +2029,7 @@ void API_EXPORTED libusb_unlock_event_waiters(libusb_context *ctx)
  * \returns \ref LIBUSB_ERROR_INVALID_PARAM if timeval is invalid
  * \ref libusb_mtasync
  */
-int API_EXPORTED libusb_wait_for_event(libusb_context *ctx, struct timeval *tv)
+int API_EXPORTED libusb_wait_for_event(libusb_context *ctx, struct timeval *tv) REQUIRES(ctx->event_waiters_lock)
 {
 	int r;
 
@@ -2210,7 +2210,7 @@ static int handle_timer_trigger(struct libusb_context *ctx)
 
 /* do the actual event handling. assumes that no other thread is concurrently
  * doing the same thing. */
-static int handle_events(struct libusb_context *ctx, struct timeval *tv)
+static int handle_events(struct libusb_context *ctx, struct timeval *tv) REQUIRES(ctx->events_lock)
 {
 	struct usbi_reported_events reported_events;
 	int r, timeout_ms;
@@ -2494,7 +2494,7 @@ int API_EXPORTED libusb_handle_events_completed(libusb_context *ctx,
  * \ref libusb_mtasync
  */
 int API_EXPORTED libusb_handle_events_locked(libusb_context *ctx,
-	struct timeval *tv)
+	struct timeval *tv) REQUIRES(ctx->events_lock)
 {
 	int r;
 	struct timeval poll_timeout;
@@ -2669,7 +2669,7 @@ void API_EXPORTED libusb_set_pollfd_notifiers(libusb_context *ctx,
  * Interrupt the iteration of the event handling thread, so that it picks
  * up the event source change. Callers of this function must hold the event_data_lock.
  */
-static void usbi_event_source_notification(struct libusb_context *ctx)
+static void usbi_event_source_notification(struct libusb_context *ctx) REQUIRES(ctx->event_data_lock)
 {
 	unsigned int event_flags;
 
@@ -2816,7 +2816,7 @@ void API_EXPORTED libusb_free_pollfds(const struct libusb_pollfd **pollfds)
  * device. This function ensures transfers get cancelled appropriately.
  * Callers of this function must hold the events_lock.
  */
-void usbi_handle_disconnect(struct libusb_context *ctx, struct libusb_device_handle *dev_handle)
+void usbi_handle_disconnect(struct libusb_context *ctx, struct libusb_device_handle *dev_handle) REQUIRES(ctx->events_lock)
 {
 	struct usbi_transfer *cur;
 	struct usbi_transfer *to_cancel;

--- a/libusb/libusbi.h
+++ b/libusb/libusbi.h
@@ -47,6 +47,8 @@
 #define static_assert(cond, msg) _Static_assert(cond, msg)
 #endif
 
+#include "clang_thread_safety.h"
+
 #ifdef NDEBUG
 #define ASSERT_EQ(expression, value)	(void)expression
 #define ASSERT_NE(expression, value)	(void)expression
@@ -373,17 +375,18 @@ struct libusb_context {
 #endif
 
 	usbi_mutex_t usb_devs_lock;
-	struct list_head usb_devs;
+	struct list_head usb_devs GUARDED_BY(usb_devs_lock);
 
 	/* A list of open handles. Backends are free to traverse this if required.
 	 */
 	usbi_mutex_t open_devs_lock;
-	struct list_head open_devs;
+	struct list_head open_devs GUARDED_BY(open_devs_lock);
 
 	/* A list of registered hotplug callbacks */
 	usbi_mutex_t hotplug_cbs_lock;
-	struct list_head hotplug_cbs;
-	libusb_hotplug_callback_handle next_hotplug_cb_handle;
+	struct list_head hotplug_cbs GUARDED_BY(hotplug_cbs_lock);
+	struct list_head* dummy GUARDED_BY(hotplug_cbs_lock);
+	libusb_hotplug_callback_handle next_hotplug_cb_handle GUARDED_BY(hotplug_cbs_lock);
 
 	/* A flag to indicate that the context is ready for hotplug notifications */
 	usbi_atomic_t hotplug_ready;
@@ -395,7 +398,7 @@ struct libusb_context {
 	 * expiration. URBs to timeout the soonest are placed at the beginning of
 	 * the list, URBs that will time out later are placed after, and urbs with
 	 * infinite timeout are always placed at the very end. */
-	struct list_head flying_transfers;
+	struct list_head flying_transfers GUARDED_BY(flying_transfers_lock);
 
 #if !defined(PLATFORM_WINDOWS)
 	/* user callbacks for pollfd changes */
@@ -408,7 +411,7 @@ struct libusb_context {
 	usbi_mutex_t events_lock;
 
 	/* used to see if there is an active thread doing event handling */
-	int event_handler_active;
+	int event_handler_active GUARDED_BY(events_lock);
 
 	/* A thread-local storage key to track which thread is performing event
 	 * handling */
@@ -424,18 +427,18 @@ struct libusb_context {
 
 	/* A bitmask of flags that are set to indicate specific events that need to
 	 * be handled. Protected by event_data_lock. */
-	unsigned int event_flags;
+	unsigned int event_flags GUARDED_BY(event_data_lock);
 
 	/* A counter that is set when we want to interrupt and prevent event handling,
 	 * in order to safely close a device. Protected by event_data_lock. */
-	unsigned int device_close;
+	unsigned int device_close GUARDED_BY(event_data_lock);
 
 	/* A list of currently active event sources. Protected by event_data_lock. */
-	struct list_head event_sources;
+	struct list_head event_sources GUARDED_BY(event_data_lock);
 
 	/* A list of event sources that have been removed since the last time
 	 * event sources were waited on. Protected by event_data_lock. */
-	struct list_head removed_event_sources;
+	struct list_head removed_event_sources GUARDED_BY(event_data_lock);
 
 	/* A pointer and count to platform-specific data used for monitoring event
 	 * sources. Only accessed during event handling. */
@@ -443,10 +446,10 @@ struct libusb_context {
 	unsigned int event_data_cnt;
 
 	/* A list of pending hotplug messages. Protected by event_data_lock. */
-	struct list_head hotplug_msgs;
+	struct list_head hotplug_msgs GUARDED_BY(event_data_lock);
 
 	/* A list of pending completed transfers. Protected by event_data_lock. */
-	struct list_head completed_transfers;
+	struct list_head completed_transfers GUARDED_BY(event_data_lock);
 
 	struct list_head list;
 };
@@ -455,7 +458,7 @@ extern struct libusb_context *usbi_default_context;
 extern struct libusb_context *usbi_fallback_context;
 
 extern usbi_mutex_static_t active_contexts_lock;
-extern struct list_head active_contexts_list;
+extern struct list_head active_contexts_list GUARDED_BY(active_contexts_lock);
 
 static inline struct libusb_context *usbi_get_context(struct libusb_context *ctx)
 {
@@ -533,7 +536,7 @@ struct libusb_device {
 struct libusb_device_handle {
 	/* lock protects claimed_interfaces */
 	usbi_mutex_t lock;
-	unsigned long claimed_interfaces;
+	unsigned long claimed_interfaces GUARDED_BY(lock);
 
 	struct list_head list;
 	struct libusb_device *dev;
@@ -604,7 +607,7 @@ struct usbi_transfer {
 	struct timespec timeout;
 	int transferred;
 	uint32_t stream_id;
-	uint32_t state_flags;   /* Protected by usbi_transfer->lock */
+	uint32_t state_flags GUARDED_BY(lock);   /* Protected by usbi_transfer->lock */
 	uint32_t timeout_flags; /* Protected by the flying_transfers_lock */
 
 	/* The device reference is held until destruction for logging

--- a/libusb/os/darwin_usb.c
+++ b/libusb/os/darwin_usb.c
@@ -66,8 +66,8 @@ static CFRunLoopRef libusb_darwin_acfl = NULL; /* event cf loop */
 static CFRunLoopSourceRef libusb_darwin_acfls = NULL; /* shutdown signal for event cf loop */
 
 static usbi_mutex_t darwin_cached_devices_mutex = USBI_MUTEX_INITIALIZER;
-static struct list_head darwin_cached_devices;
-static int init_count;
+static struct list_head darwin_cached_devices GUARDED_BY(darwin_cached_devices_mutex);
+static int init_count GUARDED_BY(darwin_cached_devices_mutex);
 
 static const char *darwin_device_class = "IOUSBDevice";
 
@@ -82,7 +82,7 @@ bool libusb_testonly_clear_running_version_cache __attribute__ ((visibility ("hi
 static pthread_t libusb_darwin_at;
 
 /* protected by libusb_darwin_at_mutex */
-static bool libusb_darwin_at_started;
+static bool libusb_darwin_at_started GUARDED_BY(libusb_darwin_at_mutex);
 
 static void darwin_exit(struct libusb_context *ctx);
 static int darwin_get_config_descriptor(struct libusb_device *dev, uint8_t config_index, void *buffer, size_t len);
@@ -448,7 +448,7 @@ uint32_t get_running_version(void) {
 }
 
 /* this function must be called with the darwin_cached_devices_mutex held */
-static void darwin_deref_cached_device(struct darwin_cached_device *cached_dev) {
+static void darwin_deref_cached_device(struct darwin_cached_device *cached_dev) REQUIRES(darwin_cached_devices_mutex) {
   cached_dev->refcount--;
   /* free the device and remove it from the cache */
   if (0 == cached_dev->refcount) {
@@ -463,7 +463,7 @@ static void darwin_deref_cached_device(struct darwin_cached_device *cached_dev) 
   }
 }
 
-static void darwin_ref_cached_device(struct darwin_cached_device *cached_dev) {
+static void darwin_ref_cached_device(struct darwin_cached_device *cached_dev) REQUIRES(darwin_cached_devices_mutex) {
   cached_dev->refcount++;
 }
 
@@ -622,7 +622,7 @@ static enum libusb_error darwin_device_from_service (struct libusb_context *ctx,
   return LIBUSB_SUCCESS;
 }
 
-static void darwin_devices_attached (void *ptr, io_iterator_t add_devices) {
+static void darwin_devices_attached (void *ptr, io_iterator_t add_devices) EXCLUDES(active_contexts_lock) {
   UNUSED(ptr);
   struct darwin_cached_device *cached_device;
   UInt64 old_session_id;
@@ -653,7 +653,7 @@ static void darwin_devices_attached (void *ptr, io_iterator_t add_devices) {
   usbi_mutex_unlock(&active_contexts_lock);
 }
 
-static void darwin_devices_detached (void *ptr, io_iterator_t rem_devices) {
+static void darwin_devices_detached (void *ptr, io_iterator_t rem_devices) EXCLUDES(darwin_cached_devices_mutex) EXCLUDES(active_contexts_lock) {
   UNUSED(ptr);
   struct libusb_device *dev = NULL;
   struct libusb_context *ctx;
@@ -740,7 +740,7 @@ static void darwin_clear_iterator (io_iterator_t iter) {
     IOObjectRelease (device);
 }
 
-static void darwin_fail_startup(void) {
+static void darwin_fail_startup(void) EXCLUDES(libusb_darwin_at_mutex) {
   usbi_mutex_lock (&libusb_darwin_at_mutex);
   libusb_darwin_acfl = LIBUSB_DARWIN_STARTUP_FAILURE;
   usbi_cond_signal (&libusb_darwin_at_cond);
@@ -748,7 +748,7 @@ static void darwin_fail_startup(void) {
   pthread_exit (NULL);
 }
 
-static void *darwin_event_thread_main (void *arg0) {
+static void *darwin_event_thread_main (void *arg0) EXCLUDES(libusb_darwin_at_mutex) {
   UNUSED(arg0);
   IOReturn kresult;
   CFRunLoopRef runloop;
@@ -866,7 +866,7 @@ static void *darwin_event_thread_main (void *arg0) {
 }
 
 /* cleanup function to destroy cached devices. must be called with a lock on darwin_cached_devices_mutex */
-static void darwin_cleanup_devices(void) {
+static void darwin_cleanup_devices(void) REQUIRES(darwin_cached_devices_mutex) {
   struct darwin_cached_device *dev, *next;
 
   list_for_each_entry_safe(dev, next, &darwin_cached_devices, list, struct darwin_cached_device) {
@@ -878,7 +878,7 @@ static void darwin_cleanup_devices(void) {
 }
 
 /* must be called with a lock on darwin_cached_devices_mutex */
-static int darwin_first_time_init(void) {
+static int darwin_first_time_init(void) REQUIRES(darwin_cached_devices_mutex) {
   if (NULL == darwin_cached_devices.next) {
     list_init (&darwin_cached_devices);
   }
@@ -919,7 +919,7 @@ static int darwin_first_time_init(void) {
   return rc;
 }
 
-static int darwin_init_context(struct libusb_context *ctx) {
+static int darwin_init_context(struct libusb_context *ctx) EXCLUDES(darwin_cached_devices_mutex) {
   usbi_mutex_lock(&darwin_cached_devices_mutex);
 
   bool first_init = (1 == ++init_count);
@@ -946,7 +946,7 @@ static int darwin_init(struct libusb_context *ctx) {
   return rc;
 }
 
-static void darwin_exit (struct libusb_context *ctx) {
+static void darwin_exit (struct libusb_context *ctx) EXCLUDES(darwin_cached_devices_mutex) {
   UNUSED(ctx);
   bool last_exit = false;
 
@@ -1361,7 +1361,7 @@ static bool get_device_parent_sessionID(io_service_t service, UInt64 *parent_ses
 }
 
 static enum libusb_error darwin_get_cached_device(struct libusb_context *ctx, io_service_t service, struct darwin_cached_device **cached_out,
-                                                  UInt64 *old_session_id) {
+                                                  UInt64 *old_session_id) EXCLUDES(darwin_cached_devices_mutex) {
   struct darwin_cached_device *new_device;
   UInt64 sessionID = 0, parent_sessionID = 0;
   UInt32 locationID = 0;
@@ -1647,7 +1647,7 @@ static int darwin_open (struct libusb_device_handle *dev_handle) {
   return 0;
 }
 
-static void darwin_close (struct libusb_device_handle *dev_handle) {
+static void darwin_close (struct libusb_device_handle *dev_handle) REQUIRES(dev_handle->lock) {
   struct darwin_device_handle_priv *priv = (struct darwin_device_handle_priv *)usbi_get_device_handle_priv(dev_handle);
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   IOReturn kresult;
@@ -1699,7 +1699,7 @@ static int darwin_get_configuration(struct libusb_device_handle *dev_handle, uin
   return LIBUSB_SUCCESS;
 }
 
-static int darwin_set_configuration(struct libusb_device_handle *dev_handle, int config) {
+static int darwin_set_configuration(struct libusb_device_handle *dev_handle, int config) REQUIRES(dev_handle->lock) {
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   IOReturn kresult;
   uint8_t i;
@@ -1843,7 +1843,7 @@ static int get_endpoints (struct libusb_device_handle *dev_handle, uint8_t iface
   return LIBUSB_SUCCESS;
 }
 
-static int darwin_claim_interface(struct libusb_device_handle *dev_handle, uint8_t iface) {
+static int darwin_claim_interface(struct libusb_device_handle *dev_handle, uint8_t iface) REQUIRES(dev_handle->lock) {
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   struct darwin_device_handle_priv *priv = (struct darwin_device_handle_priv *)usbi_get_device_handle_priv(dev_handle);
   io_service_t          usbInterface = IO_OBJECT_NULL;
@@ -1954,7 +1954,7 @@ static int darwin_claim_interface(struct libusb_device_handle *dev_handle, uint8
   return LIBUSB_SUCCESS;
 }
 
-static int darwin_release_interface(struct libusb_device_handle *dev_handle, uint8_t iface) {
+static int darwin_release_interface(struct libusb_device_handle *dev_handle, uint8_t iface) REQUIRES(dev_handle->lock) {
   struct darwin_device_handle_priv *priv = (struct darwin_device_handle_priv *)usbi_get_device_handle_priv(dev_handle);
   IOReturn kresult;
 
@@ -2014,7 +2014,7 @@ static int check_alt_setting_and_clear_halt(struct libusb_device_handle *dev_han
   return LIBUSB_SUCCESS;
 }
 
-static int darwin_set_interface_altsetting(struct libusb_device_handle *dev_handle, uint8_t iface, uint8_t altsetting) {
+static int darwin_set_interface_altsetting(struct libusb_device_handle *dev_handle, uint8_t iface, uint8_t altsetting) REQUIRES(dev_handle->lock) {
   struct darwin_device_handle_priv *priv = (struct darwin_device_handle_priv *)usbi_get_device_handle_priv(dev_handle);
   IOReturn kresult;
   int ret;
@@ -2064,7 +2064,7 @@ static int darwin_set_interface_altsetting(struct libusb_device_handle *dev_hand
   return ret;
 }
 
-static int darwin_clear_halt(struct libusb_device_handle *dev_handle, unsigned char endpoint) {
+static int darwin_clear_halt(struct libusb_device_handle *dev_handle, unsigned char endpoint) EXCLUDES(dev_handle->lock) {
   /* current interface */
   struct darwin_interface *cInterface;
   IOReturn kresult;
@@ -2086,7 +2086,7 @@ static int darwin_clear_halt(struct libusb_device_handle *dev_handle, unsigned c
 }
 
 static enum libusb_error darwin_restore_state (struct libusb_device_handle *dev_handle, uint8_t active_config,
-                                               unsigned long claimed_interfaces) {
+                                               unsigned long claimed_interfaces) REQUIRES(dev_handle->lock) {
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   struct darwin_device_handle_priv *priv = (struct darwin_device_handle_priv *)usbi_get_device_handle_priv(dev_handle);
   int open_count = dpriv->open_count;
@@ -2147,7 +2147,7 @@ static enum libusb_error darwin_restore_state (struct libusb_device_handle *dev_
   return LIBUSB_SUCCESS;
 }
 
-static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, bool capture) {
+static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, bool capture) REQUIRES(dev_handle->lock) {
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   unsigned long claimed_interfaces = dev_handle->claimed_interfaces;
   uint8_t active_config = dpriv->active_config;
@@ -2253,7 +2253,7 @@ static int darwin_reenumerate_device (struct libusb_device_handle *dev_handle, b
   return darwin_restore_state (dev_handle, active_config, claimed_interfaces);
 }
 
-static int darwin_reset_device (struct libusb_device_handle *dev_handle) {
+static int darwin_reset_device (struct libusb_device_handle *dev_handle) REQUIRES(dev_handle->lock) {
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   IOReturn kresult;
   int ret;
@@ -2337,7 +2337,7 @@ static int darwin_kernel_driver_active(struct libusb_device_handle *dev_handle, 
   return 0;
 }
 
-static void darwin_destroy_device(struct libusb_device *dev) {
+static void darwin_destroy_device(struct libusb_device *dev) EXCLUDES(darwin_cached_devices_mutex) {
   struct darwin_device_priv *dpriv = (struct darwin_device_priv *)usbi_get_device_priv(dev);
 
   if (dpriv->dev) {
@@ -2349,7 +2349,7 @@ static void darwin_destroy_device(struct libusb_device *dev) {
   }
 }
 
-static int submit_bulk_transfer(struct usbi_transfer *itransfer) {
+static int submit_bulk_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 
   IOReturn               ret;
@@ -2406,7 +2406,7 @@ static int submit_bulk_transfer(struct usbi_transfer *itransfer) {
 }
 
 #if MAX_INTERFACE_VERSION >= 550
-static int submit_stream_transfer(struct usbi_transfer *itransfer) {
+static int submit_stream_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
   struct darwin_interface *cInterface;
   uint8_t pipeRef;
@@ -2443,7 +2443,7 @@ static int submit_stream_transfer(struct usbi_transfer *itransfer) {
 }
 #endif
 
-static int submit_iso_transfer(struct usbi_transfer *itransfer) {
+static int submit_iso_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
   struct darwin_transfer_priv *tpriv = (struct darwin_transfer_priv *)usbi_get_transfer_priv(itransfer);
 
@@ -2535,7 +2535,7 @@ static int submit_iso_transfer(struct usbi_transfer *itransfer) {
   return darwin_to_libusb (kresult);
 }
 
-static int submit_control_transfer(struct usbi_transfer *itransfer) {
+static int submit_control_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
   struct libusb_control_setup *setup = (struct libusb_control_setup *) transfer->buffer;
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(transfer->dev_handle->dev);
@@ -2583,7 +2583,7 @@ static int submit_control_transfer(struct usbi_transfer *itransfer) {
   return darwin_to_libusb (kresult);
 }
 
-static int darwin_submit_transfer(struct usbi_transfer *itransfer) {
+static int darwin_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 
   switch (transfer->type) {
@@ -2607,7 +2607,7 @@ static int darwin_submit_transfer(struct usbi_transfer *itransfer) {
   }
 }
 
-static int cancel_control_transfer(struct usbi_transfer *itransfer) {
+static int cancel_control_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(transfer->dev_handle->dev);
   IOReturn kresult;
@@ -2623,7 +2623,7 @@ static int cancel_control_transfer(struct usbi_transfer *itransfer) {
   return darwin_to_libusb (kresult);
 }
 
-static int darwin_abort_transfers (struct usbi_transfer *itransfer) {
+static int darwin_abort_transfers (struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(transfer->dev_handle->dev);
   struct darwin_interface *cInterface;
@@ -2665,7 +2665,7 @@ static int darwin_abort_transfers (struct usbi_transfer *itransfer) {
   return darwin_to_libusb (kresult);
 }
 
-static int darwin_cancel_transfer(struct usbi_transfer *itransfer) {
+static int darwin_cancel_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock) {
   struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 
   switch (transfer->type) {
@@ -2895,7 +2895,7 @@ static bool darwin_has_capture_entitlements (void) {
   return entitled;
 }
 
-static int darwin_reload_device (struct libusb_device_handle *dev_handle) {
+static int darwin_reload_device (struct libusb_device_handle *dev_handle) EXCLUDES(darwin_cached_devices_mutex) {
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   enum libusb_error err;
 
@@ -2909,7 +2909,7 @@ static int darwin_reload_device (struct libusb_device_handle *dev_handle) {
 
 /* On macOS, we capture an entire device at once, not individual interfaces. */
 
-static int darwin_detach_kernel_driver (struct libusb_device_handle *dev_handle, uint8_t interface) {
+static int darwin_detach_kernel_driver (struct libusb_device_handle *dev_handle, uint8_t interface) REQUIRES(dev_handle->lock) {
   UNUSED(interface);
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
   IOReturn kresult;
@@ -2955,7 +2955,7 @@ static int darwin_detach_kernel_driver (struct libusb_device_handle *dev_handle,
 }
 
 
-static int darwin_attach_kernel_driver (struct libusb_device_handle *dev_handle, uint8_t interface) {
+static int darwin_attach_kernel_driver (struct libusb_device_handle *dev_handle, uint8_t interface) REQUIRES(dev_handle->lock) {
   UNUSED(interface);
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
 
@@ -2974,7 +2974,7 @@ static int darwin_attach_kernel_driver (struct libusb_device_handle *dev_handle,
   return darwin_reenumerate_device (dev_handle, false);
 }
 
-static int darwin_capture_claim_interface(struct libusb_device_handle *dev_handle, uint8_t iface) {
+static int darwin_capture_claim_interface(struct libusb_device_handle *dev_handle, uint8_t iface) REQUIRES(dev_handle->lock) {
   int ret;
   if (dev_handle->auto_detach_kernel_driver && darwin_kernel_driver_active(dev_handle, iface)) {
     ret = darwin_detach_kernel_driver (dev_handle, iface);
@@ -2986,7 +2986,7 @@ static int darwin_capture_claim_interface(struct libusb_device_handle *dev_handl
   return darwin_claim_interface (dev_handle, iface);
 }
 
-static int darwin_capture_release_interface(struct libusb_device_handle *dev_handle, uint8_t iface) {
+static int darwin_capture_release_interface(struct libusb_device_handle *dev_handle, uint8_t iface) REQUIRES(dev_handle->lock) {
   int ret;
   struct darwin_cached_device *dpriv = DARWIN_CACHED_DEVICE(dev_handle->dev);
 

--- a/libusb/os/emscripten_webusb.cpp
+++ b/libusb/os/emscripten_webusb.cpp
@@ -764,7 +764,7 @@ void em_destroy_device(libusb_device* dev) {
 	WebUsbDevicePtr(dev).free();
 }
 
-int em_submit_transfer(usbi_transfer* itransfer) {
+int em_submit_transfer(usbi_transfer* itransfer) REQUIRES(itransfer->lock) {
 	return runOnMain([itransfer] {
 		auto transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 		auto& web_usb_device = WebUsbDevicePtr(transfer->dev_handle)

--- a/libusb/os/haiku_usb_raw.cpp
+++ b/libusb/os/haiku_usb_raw.cpp
@@ -140,7 +140,7 @@ haiku_release_interface(struct libusb_device_handle *dev_handle, uint8_t interfa
 }
 
 static int
-haiku_submit_transfer(struct usbi_transfer *itransfer)
+haiku_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	struct libusb_transfer *fLibusbTransfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 	USBDeviceHandle *fDeviceHandle = *((USBDeviceHandle **)usbi_get_device_handle_priv(fLibusbTransfer->dev_handle));

--- a/libusb/os/linux_usbfs.c
+++ b/libusb/os/linux_usbfs.c
@@ -2326,7 +2326,7 @@ static int submit_control_transfer(struct usbi_transfer *itransfer)
 	return 0;
 }
 
-static int op_submit_transfer(struct usbi_transfer *itransfer)
+static int op_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	struct libusb_transfer *transfer =
 		USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);

--- a/libusb/os/netbsd_usb.c
+++ b/libusb/os/netbsd_usb.c
@@ -387,7 +387,7 @@ netbsd_destroy_device(struct libusb_device *dev)
 }
 
 int
-netbsd_submit_transfer(struct usbi_transfer *itransfer)
+netbsd_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	struct libusb_transfer *transfer;
 	int err = 0;

--- a/libusb/os/null_usb.c
+++ b/libusb/os/null_usb.c
@@ -85,7 +85,7 @@ null_clear_halt(struct libusb_device_handle *handle, unsigned char endpoint)
 }
 
 static int
-null_submit_transfer(struct usbi_transfer *itransfer)
+null_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	return LIBUSB_ERROR_NOT_SUPPORTED;
 }

--- a/libusb/os/openbsd_usb.c
+++ b/libusb/os/openbsd_usb.c
@@ -425,7 +425,7 @@ obsd_destroy_device(struct libusb_device *dev)
 }
 
 int
-obsd_submit_transfer(struct usbi_transfer *itransfer)
+obsd_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	struct libusb_transfer *transfer;
 	int err = 0;

--- a/libusb/os/sunos_usb.c
+++ b/libusb/os/sunos_usb.c
@@ -1384,7 +1384,7 @@ sunos_destroy_device(struct libusb_device *dev)
 }
 
 int
-sunos_submit_transfer(struct usbi_transfer *itransfer)
+sunos_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	struct	libusb_transfer *transfer;
 	struct	libusb_device_handle *hdl;

--- a/libusb/os/threads_posix.c
+++ b/libusb/os/threads_posix.c
@@ -56,7 +56,7 @@ void usbi_cond_init(usbi_cond_t *cond)
 }
 
 int usbi_cond_timedwait(usbi_cond_t *cond,
-	usbi_mutex_t *mutex, const struct timeval *tv)
+	usbi_mutex_t *mutex, const struct timeval *tv) REQUIRES(*mutex)
 {
 	struct timespec timeout;
 	int r;

--- a/libusb/os/threads_posix.h
+++ b/libusb/os/threads_posix.h
@@ -29,35 +29,35 @@
 #define PTHREAD_CHECK(expression)	ASSERT_EQ(expression, 0)
 
 #define USBI_MUTEX_INITIALIZER	PTHREAD_MUTEX_INITIALIZER
-typedef pthread_mutex_t usbi_mutex_static_t;
-static inline void usbi_mutex_static_lock(usbi_mutex_static_t *mutex)
+CAPABILITY("mutex") typedef pthread_mutex_t usbi_mutex_static_t;
+static inline void usbi_mutex_static_lock(usbi_mutex_static_t *mutex) ACQUIRE(*mutex) NO_THREAD_SAFETY_ANALYSIS
 {
 	PTHREAD_CHECK(pthread_mutex_lock(mutex));
 }
-static inline void usbi_mutex_static_unlock(usbi_mutex_static_t *mutex)
+static inline void usbi_mutex_static_unlock(usbi_mutex_static_t *mutex) RELEASE(*mutex) NO_THREAD_SAFETY_ANALYSIS
 {
 	PTHREAD_CHECK(pthread_mutex_unlock(mutex));
 }
 
-typedef pthread_mutex_t usbi_mutex_t;
-static inline void usbi_mutex_init(usbi_mutex_t *mutex)
+CAPABILITY("mutex") typedef pthread_mutex_t usbi_mutex_t;
+static inline void usbi_mutex_init(usbi_mutex_t *mutex) EXCLUDES(*mutex)
 {
 	PTHREAD_CHECK(pthread_mutex_init(mutex, NULL));
 }
-static inline void usbi_mutex_lock(usbi_mutex_t *mutex)
+static inline void usbi_mutex_lock(usbi_mutex_t *mutex) ACQUIRE(*mutex) NO_THREAD_SAFETY_ANALYSIS
 {
 	PTHREAD_CHECK(pthread_mutex_lock(mutex));
 }
-static inline void usbi_mutex_unlock(usbi_mutex_t *mutex)
+static inline void usbi_mutex_unlock(usbi_mutex_t *mutex) RELEASE(*mutex) NO_THREAD_SAFETY_ANALYSIS
 {
 	PTHREAD_CHECK(pthread_mutex_unlock(mutex));
 }
-static inline int usbi_mutex_trylock(usbi_mutex_t *mutex)
+static inline int usbi_mutex_trylock(usbi_mutex_t *mutex) TRY_ACQUIRE(1, *mutex)
 {
 	int mutexIsLocked = pthread_mutex_trylock(mutex) == 0;
 	return mutexIsLocked;
 }
-static inline void usbi_mutex_destroy(usbi_mutex_t *mutex)
+static inline void usbi_mutex_destroy(usbi_mutex_t *mutex) EXCLUDES(*mutex)
 {
 	PTHREAD_CHECK(pthread_mutex_destroy(mutex));
 }
@@ -65,7 +65,7 @@ static inline void usbi_mutex_destroy(usbi_mutex_t *mutex)
 #define USBI_COND_INITIALIZER	PTHREAD_COND_INITIALIZER
 typedef pthread_cond_t usbi_cond_t;
 void usbi_cond_init(usbi_cond_t *cond);
-static inline void usbi_cond_wait(usbi_cond_t *cond, usbi_mutex_t *mutex)
+static inline void usbi_cond_wait(usbi_cond_t *cond, usbi_mutex_t *mutex) REQUIRES(*mutex)
 {
 	PTHREAD_CHECK(pthread_cond_wait(cond, mutex));
 }

--- a/libusb/os/windows_common.c
+++ b/libusb/os/windows_common.c
@@ -784,7 +784,7 @@ static int windows_get_max_raw_io_transfer_size(struct libusb_device_handle *dev
 	return LIBUSB_ERROR_NOT_SUPPORTED;
 }
 
-static int windows_submit_transfer(struct usbi_transfer *itransfer)
+static int windows_submit_transfer(struct usbi_transfer *itransfer) REQUIRES(itransfer->lock)
 {
 	struct libusb_transfer *transfer = USBI_TRANSFER_TO_LIBUSB_TRANSFER(itransfer);
 	struct libusb_device_handle *dev_handle = transfer->dev_handle;
@@ -859,7 +859,7 @@ static int windows_cancel_transfer(struct usbi_transfer *itransfer)
 	return LIBUSB_ERROR_NOT_SUPPORTED;
 }
 
-static int windows_handle_transfer_completion(struct usbi_transfer *itransfer)
+static int windows_handle_transfer_completion(struct usbi_transfer *itransfer) EXCLUDES(itransfer->lock)
 {
 	struct libusb_context *ctx = ITRANSFER_CTX(itransfer);
 	struct windows_context_priv *priv = usbi_get_context_priv(ctx);

